### PR TITLE
Refactor inference optimizations

### PR DESCRIPTION
**Pull request by xingniu**
This PR refactors inference optimizations:
- Make the `inference_only` mode switchable.
- Simplify inference optimizations by
  (1) using `eval()` to disable dropout instead of explicitly setting dropout modules to None;
  (2) always using default value `inplace=False` for activation modules.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [ ] ~~Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?~~
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [ ] ~~You have considered writing a test~~
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/model.py [hunk 1]
```diff
@@ -189,6 +191,19 @@ def __init__(self,
 
         self.knn : Optional[layers.KNN] = None
 
+    def set_inference_only(self, inference_only: bool):
+        """
+        Turn inference_only/eval mode on or off.
+        """
+        if self.inference_only != inference_only:
```
Comment (mjdenkowski) [inline sockeye/model.py:198]: Is this check needed? Does setting `inference_only` to the same value multiple times have any negative effects?

File: sockeye/model.py [hunk 2]
```diff
@@ -144,10 +144,12 @@ def __init__(self,
                                                vocab_size=self.config.vocab_target_size,
                                                weight=output_weight,
                                                dtype=self.dtype)
+        self.output_layer_module_cached = self.output_layer
+        self.output_layer_script_cached = pt.jit.script(self.output_layer_module_cached)
         if self.inference_only:
             # Running this layer scripted with a newly initialized model can
             # cause an overflow error.
-            self.output_layer = pt.jit.script(self.output_layer)
+            self.output_layer = self.output_layer_script_cached
```
Comment (mjdenkowski) [inline sockeye/model.py:152]: Could this be replaced with a call to `self.set_inference_only`?

File: sockeye/model.py [hunk 3]
```diff
@@ -189,6 +191,19 @@ def __init__(self,
...
+        Turn inference_only/eval mode on or off.
+        """
+        if self.inference_only != inference_only:
+            self.inference_only = inference_only
+            if self.inference_only:
+                self.output_layer = self.output_layer_script_cached
+            else:
+                self.output_layer = self.output_layer_module_cached
+            self.decoder.set_inference_only(self.inference_only)
+            self.train(not inference_only)
```
Comment (mjdenkowski) [inline sockeye/model.py:205]: Do we need to set `self.train` here? Could we say that `self.set_inference_only` controls Sockeye-level optimizations (cell state transform and output layer) while `self.train` controls PyTorch-level optimizations (dropout)? This would decouple the two settings to avoid potential unintended side effects when users change one or the other.

**Review by mjdenkowski — COMMENTED** [review]
Thanks for working on this! I've added a few comments.

File: sockeye/layers.py [hunk 4]
```diff
@@ -812,6 +823,13 @@ def __init__(self,
 
         self.relu = pt.nn.ReLU(inplace=False)  # inplace=False because we need to non-activated data as well
 
+    def set_inference_only(self, inference_only: bool):
+        """
+        Set inference_only.
+        """
```
Comment (mjdenkowski) [inline sockeye/layers.py:829]: For consistency, we could add:
```
self.inference_only = inference_only
```

File: sockeye/transformer.py [hunk 5]
```diff
@@ -203,6 +203,13 @@ def __init__(self,
         if config.use_lhuc:
             self.lhuc = sockeye.layers.LHUC(config.model_size, dtype=dtype)
 
+    def set_inference_only(self, inference_only: bool):
+        """
+        Set inference_only.
+        """
```
Comment (mjdenkowski) [inline sockeye/transformer.py:209]: For consistency, we could add:
```
self.inference_only = inference_only
```

**Review by mjdenkowski — COMMENTED** [review]
Thanks Xing! Can you make another pass to double check that any class that has a `set_inference_only` method also sets `self.inference_only` in both `set_inference_only` and `__init__`? The idea is that any class that supports switching in/out of inference mode can always be queried for its current state.

**Review by mjdenkowski — APPROVED** [review]
Looks good!
